Added an AI policy (#29)
strimzi/governance: AI_POLICY.md

# AI Contribution Policy for the Strimzi Project

## Overview

Contributors may use AI tools, such as Large Language Models (LLMs) and code assistants, when contributing to Strimzi.
As with any development tool, the contributor is responsible for the quality of the result and for understanding what they submit.

**You are the contributor.** When you submit a Pull Request (PR) to Strimzi, you certify the contribution as your own work.
AI-generated or AI-assisted content does not change this responsibility.

## Requirements

> ### Understanding and Responsibility

Reviewer: Should we add something like that the user is responsible to replying to any comments / feedback themselves and not just let AI do it for them?

Author: Isn't it covered around line 60 "Engage authentically with reviewers."?

Reviewer: I do not think it is the same. That point says _you can use AI to talk for you as long as you think you understand it_. (which frankly many people would think even if they have no clue)

Author: So are you suggesting that users should not use AI assistant at all to provide comments? I think that using AI assistant to craft or improve something that you want to say is not that bad and should not be forbidden.

Reviewer: I'm happy with people using things to translate, improve grammar, etc. I'm even happy with people asking their AI about problems raised in the comments. I'm not hjappy with people telling their AI to reply to the commnets raised on their PRs. And I do not think that is clear from that section.

Author: @scholzj can you please suggest something better (or what you have in mind) that we could have here?

Reviewer: Kubernetes states this:

```
You should be able to answer reviewers' questions yourself, without recourse to AI. In particular, do not waste the time of other contributors by being a proxy between reviewers and an AI.
```

Maybe we could add it to the line 15?

```
* **You must understand your contribution.** Do not submit code, documentation, or other content that you do not fully understand. You should be able to explain to reviewers what your contribution does, how it works, and how you verified it. You should be able to answer reviewers' questions yourself, without recourse to AI. In particular, do not waste the time of other contributors by being a proxy between reviewers and an AI.
```

* **You must understand your contribution.** Do not submit code, documentation, or other content that you do not fully understand. You should be able to explain to reviewers what your contribution does, how it works, and how you verified it.
* **Review all AI-generated content.** Before submitting, verify that the contribution is correct, tested, and follows the project's contributing guidelines.

> * **Use AI as a learning tool.** New contributors could use AI tools to understand the codebase, patterns and best practices. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.

Reviewer 2: ```suggestion
* **Use AI as a learning tool.** New contributors could use AI tools to understand the codebase, patterns, and best practices. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.
```

> * **Use AI as a learning tool.** New contributors could use AI tools to understand the codebase, patterns and best practices. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.

Reviewer: While I think this is a useful part of using AI, we should not be prescribing what the users should do with the AI and how they should use it. We should limit ourself to how their use of AI interacts with Strimzi. So I do not think this belongs here.

Author: While it's not a direct interaction with Strimzi in the sense that AI is used to generate content for Strimzi, I think keeping this sentence doesn't hurt. We are just suggesting that a good understanding of Strimzi (or part of it, depending on the feature/bug you are working on) is important to work on it and the AI tools can be used to get more knowledge.

Reviewer: I do not think we are in a position to tell people how/what to learn.

Reviewer 3: `and best practices` - I think to use AI for such things we should have some AI specific md files with details what are the best practices. Otherwise each AI tool can return different outputs based on scanning the code base I am afraid. I would be very careful about it.

Author: > and best practices

I would agree on this and will remove but the rest still stands.

> I do not think we are in a position to tell people how/what to learn.

We are not telling people "how/what to learn", we are just saying that an AI assistant can be helpful to learn a new codebase.

Reviewer: Sorry, but `Use AI as a learning tool` is telling people what should they do. Leave it to them and their fantasy and focus on the AI things that do concern Strimzi.

Author: any thoughts from @strimzi/maintainers ?

Reviewer 4: I read this not as us prescribing what people should do, but more just scene setting for the kind of things we are happy with people doing, so I am happy with the wording. This could also be a space to mention things like the agents.md and claude.md files

Reviewer 2: I'm not sure that bullet is needed in this policy doc. AI or otherwise, they should familiarise with the code before submitting anything. I'd drop the bullet

Reviewer 5: Maybe this can be slightly reworded to make it sound more suggestive. For example:                                                                                                                                                               

- **Consider AI as a learning tool.** New contributors might find AI tools helpful for understanding the codebase, patterns and best practices. If you're using AI to generate code you don't fully understand, it's worth taking time to learn the underlying concepts before submitting. 

OR                                                                                                                                                                                                                                                                    
- **AI can be a helpful learning tool.** New contributors might find AI tools helpful for understanding the codebase, patterns and best practices. When using AI to generate code, consider taking time to understand the concepts behind it before submitting your contribution.

Reviewer: I think the first part of the second suggestion sounds reasonably innocent:

```
AI can be a helpful learning tool. New contributors might find AI tools helpful for understanding the codebase, patterns, and best practices.
```

But the second part is once again telling people what to do, what to use AI for, and trying to teach them some values. That is not our business. We should leave it up to them how to use it and what do they want or do not want to understand. Not every contributor wants to learn and understand Strimzi and care about it. And especially with AI, a lot of people use AI exactly because they do not want to learn something new. They often just want to get over with some task. And sometimes they have very good reasons for it.

Reviewer: I guess the part of the problem is also where it is places. In is in a section where the things are clear orders. It might look and sound very differently if you add it to some separate section at the end called _Interesting ways how to use AI with Strimzi_. Or _Topis & Tricks for using AI with Strimzi_. Not in a section called requremenets and subsection called responsiities.

Author: > We should leave it up to them how to use it and what do they want or do not want to understand. Not every contributor wants to learn and understand Strimzi and care about it. And especially with AI, a lot of people use AI exactly because they do not want to learn something new. They often just want to get over with some task. And sometimes they have very good reasons for it.

If I am using AI to generate code and I am taking fully responsability of such code as we already explain within this policy, then at least I have to understand what was generated by the AI.
Anyway, I think we are wasting already a lot of time on a single sentence which doesn't add anything bad. Despite I am convinced it's useful, I will remove it.

Reviewer: > If I am using AI to generate code and I am taking fully responsability of such code as we already explain within this policy, then at least I have to understand what was generated by the AI.

That is true, you should understand the code. But that does not mean you are required to use AI as a learning tool as this point was suggesting.

* **You own what you submit.** As the PR author, you are responsible for all changes, regardless of the tools used to create them.

> ### Disclosure

Reviewer: I disagree with the two sections - this one and the one after.
* It makes the UX hard by trying to insist on some special disclosure while forbidding the most common form of disclosure. (not to forget the DCO we need from the contributors)
* The description is completely meaningless as it is absolutely not clear what a significant or insignificant contribution. So it is not enforceable.
* While it is clear we need the DCO signed by human, I think it is not clear if a human can sign the DCO for the AI generated code. So I think this is another reason why to leave this thing completely aside.

What if we use for the disclosure a PR template?

Reviewer: I think what I'm suggesting here is inline with the Kubernetes policy:

```
If you used AI tools in preparing your PR, you must disclose this in the description of your PR. Listing AI tooling as a co-author, co-signing commits using an AI tool, or using the assisted-by, co-developed or similar commit trailer is not allowed.
```

Author: > It makes the UX hard by trying to insist on some special disclosure while forbidding the most common form of disclosure. (not to forget the DCO we need from the contributors)

I understand the UX concern, but the distinction between `Co-authored-by` and `Assisted-by` is intentional. For me, `Co-authored-by` implies joint ownership, legal standing, and shared responsibility which only apply to human collaborators who can sign the DCO.

`Assisted-by` provides transparency at commit level without positioning the AI tool as author. When someone investigates a commit months later with `git blame`, they can see AI was involved but it's not a legal contributor.

Finally, we're using SHOULD (not MUST) for `Assisted-by`, it's encouragement rather than a hard requirement. The only hard rule is the negative one: don't call AI a co-author, because it fundamentally isn't one.

> The description is completely meaningless as it is absolutely not clear what a significant or insignificant contribution. So it is not enforceable.

We have lists with some examples. Of course, it's not enforceable but I think the examples make the point.

> While it is clear we need the DCO signed by human, I think it is not clear if a human can sign the DCO for the AI generated code. So I think this is another reason why to leave this thing completely aside.

"it is not clear if a human can sign the DCO for the AI generated code" ... I think we state this already in several places. Around line 8, where we state "you are the contributor" even if the code is AI generated.  Around line 18, where we say "you own what you submit". Finally, around lines 59-60 where we mention that only humans can sigh with DCO. Can you point me to any specific places where it's not clear and we should add something more?

Reviewer: I think we should either follow the Kubernetes policy to forbid this. Or at least stop encouraging some parts of it while forbidding others. There is no real difference between coauthored and assisted TBH.

Author: > There is no real difference between coauthored and assisted TBH.

Well we are in different positions and I already explained mine. For me they are different from a "legal" point of view.

Reviewer: That is why I actually suggest to follow the lead of the main CNCF project rather then to follow my legal opinions. 😉

Reviewer 4: I agree with the comment here that regardless of having examples we shouldn't have something be required only for "significant" contributions. That is subjective and I don't see any problem with people declaring use of AI even for cases where the AI contributions is small. So I would rather we just say that any use of AI must be disclosed.

In terms of how it is disclosed I am happy with it being as part of the PR description. I think creating a distinction between the `Assisted-by` and `co-authored` by trailers is fine, so I don't have a problem with us allowing one and not the other. However, I wonder what the benefit of the trailer is given the contributor already has to put it in the PR description. For example when you do the `git blame` and see that it was assisted by AI it doesn't change who is responsible for the code, so perhaps having it disclosed only as part of the PR is simplest 🤔

Reviewer 3: > However, I wonder what the benefit of the trailer is given the contributor already has to put it in the PR description. For example when you do the git blame and see that it was assisted by AI it doesn't change who is responsible for the code, so perhaps having it disclosed only as part of the PR is simplest 

I wonder as well. The main point of all AI policies seems to be DCO or resolution that a contributor is fully responsible for the contribution (at least from my POV). The whole trailer think is how exactly useful for us? Are we going to scan every commit for that to get what actually? Don't get me wrong, if anyone want to add it. I am fine with it, I just don't understand the purpose for us (guess it is good for companies that owns the AI assistants to track their contributions, but we don't care). 

Kubernetes AI policy looks great to me - short, reasonable, and I am pretty sure well discussed by many people - https://www.kubernetes.dev/docs/guide/pull-requests/#ai-guidance

Author: > so I don't have a problem with us allowing one and not the other. However, I wonder what the benefit of the trailer is given the contributor already has to put it in the PR description

@katheris sorry but I am now confused about what's your position. We have me supporting the `Assisted-by` and Jakub(s) against it (or maybe even what @Frawless said is confusing "Don't get me wrong, if anyone want to add it. I am fine with it"). Should we all take a clear position so I can move forward to keep or remove it?

Reviewer 3: I am not fully against it, but I do not see any added value for us to having the trailers in commits. And I like the idea to keep it as much simple as possible like Kubernetes policy is.

Reviewer 4: Sorry @ppatierno I should have posed my comment as a question: Given the user will have disclosed the use of AI in the PR description, what benefit do you see in encouraging them to also include the trailer?

Author: @katheris that from an history point of you having it directly in the commit looks immediate to me than going to the PR and reading the description. Or for example, when you need to do some cherry-pick and you are not looking at the PR at all.

Reviewer 4: But once the PR is reviewed and approved and merged is there any reason for needing to know if a particular line of code or commit was written with the assistance of AI? Given that the contributor has taken ownership of the code. (Just trying to understand what I'm missing)

Author: I don't think you are missing anything. They are just different point of views. So despite the "neutral" position I guess the majority is for removing it and going the Kubernetes way. I will do it.

Reviewer 6: Sorry for being late to the party, I'm for keeping it simple as possible. I'm not that big user of AI assistants during coding. But I rather blame some real person in `git blame` than person + AI. And I'm not checking who exactly did what when I'm cherry-picking the commits. If it should be in the PR, then it doesn't need to be in the commits. That's at least my POV.

Reviewer 7: +1 for keeping it as simple as possible. I don't care about `Assisted-by` vs `Co-authored-by` I will follow here **Kubernetes** as they seem to treat them interchangeably.

* **Disclose AI usage in your PR.** If AI tools play a significant role in a contribution, note this in the PR description.

> * **Use the `Assisted-by` trailer.** Commits with substantial AI assistance SHOULD include an `Assisted-by` trailer identifying the tool and model used.

Reviewer 2: ```suggestion
* **Use the `Assisted-by` trailer.** Commits with substantial AI assistance MUST include an `Assisted-by` trailer identifying the tool and model used.
```

Reviewer 2: If we want to be categorical

Reviewer 2: what would merit _insubstantial_ against _substantial_ AI assistance? 
The way it's used? The quantity? Should we clarify?

Reviewer 3: I think `should` is correctly used. My understanding of MUST here (and correct me if I am wrong @PaulRMellor) is that it is something required instead of nice to have. We require DCOs and we state that contributor owns the contribution and sign it. Contributor is responsible for the code and it doesn't matter if the commits will have trailer or not. I mean it will be nice to have probably, but we can't enforce it anyway.

Author: Yeah exactly what Jakub said. We encourage people to have it but we cannot force them.

  
  Example commit message:
  ```
  Fix reconciliation loop for Kafka cluster scaling
  
  The operator was not properly updating the StrimziPodSet replica count
  when scaling Kafka clusters.
  
  Assisted-by: Claude Sonnet 4.5 <noreply@anthropic.com>
  ```
  

>   You can add this trailer by using:

Reviewer 2: ```suggestion
  You can add this trailer using:
```

  ```sh
  git commit -m "Your commit message" --trailer "Assisted-by: <tool>/<model>"
  ```

>   AI tools can be also configured to add this trailer automatically on commit.

Reviewer 2: ```suggestion
  AI tools can also be configured to add this trailer automatically on commit.
```

> * **When disclosure is required:** Disclosure is required when AI tools are used to generate substantial content. Using AI features similar to IDE capabilities does not require disclosure.

Reviewer 2: ```suggestion
* **When disclosure is required:** Disclosure is required when AI tools generate substantial content. Using AI features similar to IDE capabilities does not require disclosure.
```

  
  **Examples of substantial assistance requiring disclosure:**
  - Generating entire methods, classes, or modules.
  - Creating entire test suites or documentation sections.
  - Refactoring significant portions of code.
  - Generating configuration files or build scripts.
  
  **Examples NOT requiring disclosure:**
  - Autocomplete suggestions (1-2 lines).
  - Syntax corrections and formatting.

>   - Variable or method name suggestions

Reviewer 2: ```suggestion
  - Variable or method name suggestions.
```

  - Spelling checks or grammar corrections.
  - Simple code completion similar to IDE features.

### What NOT to Do

* **Do not use AI tools as co-authors.** AI tools MUST NOT be added to `Signed-off-by` or `Co-authored-by` tags in commit messages. Only humans can legally certify the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). The `Signed-off-by` tag certifies that you wrote the code or otherwise have the right to submit it under the project's license. An AI tool is not a legal entity and cannot make this certification.

> * **Engage authentically with reviewers.** When maintainers, component owners or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated replies. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.

Reviewer 2: ```suggestion
* **Engage authentically with reviewers.** When maintainers, component owners, or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated responses. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.
```

* **Do not submit large AI-generated PRs.** AI tools can generate content faster than reviewers can read it. Keep PRs focused and appropriately sized. Strimzi maintainers MAY close pull requests that are too large to review effectively or where the contributor does not appear to understand the changes.
* **Avoid verbose AI-generated content.** Trim down unnecessary verbosity in PR descriptions, commit messages, and code comments. Be clear, concise, and specific. Respect the time of maintainers and reviewers.

### Quality Standards

* **Meet the same standards.** AI-assisted contributions are reviewed to the same standard as any other contribution. Code must be correct, maintainable, well-tested, and consistent with project conventions.

> * **Ensure licensing compliance.** AI-generated content must not introduce material under licenses incompatible with the project's [Apache 2.0 License](./LICENSE).

Reviewer 2: How should contributors verify license compatibility of AI-generated content?
Do we want to give any guidance?

Reviewer: +1, I think this is not enforceable and does not set clear expectations. Maybe we should replace it by saying that the user has to sign the DCO and they have to make sure that they have `the right to submit the code they are contributing to the project`? It is not just about license, it is also about copyright, etc.

Reviewer 4: I would also be +1 for keeping this line but also adding something about the contributor needing to make sure they have the right to submit the code they are contributing

Author: I think there are a lot of things we can't enforce. We are not setting up a process with some automation doing validations for us but mostly writing a policy and people can ... break policies if they want.
I think we should keep this line. I am happy to add another one about "having the rights" but it's not clear where even because we already mention the DCO in the "What NOT to Do" section.

Reviewer 4: I think you could add it in the sentence `**You own what you submit.** As the PR author, you are responsible for all changes, regardless of the tools used to create them.` so having:

```
**You own what you submit.** As the PR author, you are responsible for all changes, regardless of the tools used to create them, and must certify that you wrote or otherwise have the right to submit the code you are contributing by signing your commits.
```

It's perhaps a little repetitive, but the existing DCO reference isn't as explicit and I don't think it's the end of the world to re-iterate things here

Reviewer 3: I would change the line to DCO style instead of licence compatibility.

* **Test your changes.** All code contributions must be tested according to the project's standards, whether AI-assisted or not.

## Consequences of Non-Compliance

Pull Requests that violate this policy may be closed. Specific cases include:

* PRs where the contributor cannot explain or defend the changes during review.
* PRs with generic, inauthentic responses to reviewer feedback that demonstrate a lack of understanding.
* Large, unfocused PRs that appear to be bulk AI-generated content.

> ## AI-Assisted Pull Request Reviews

Reviewer: Can we add that contributors should not request/do any AI-assisted reviews on the PRs? (I.e. they can do it on the code in private, but should not send any AI to comment on the Strimzi PRs). Where AI review is desired, the maintainers can request it themself.

Author: Yeah, fine with me adding it.

> Automated or AI-assisted reviews, can supplement the review process but do not replace review by maintainers or component owners.

Reviewer 2: ```suggestion
Automated or AI-assisted reviews can supplement the review process but do not replace review by maintainers or component owners.
```

* **AI reviews don't replace human review.** Automated checks and AI-generated review comments can provide useful signals (e.g., identifying security vulnerabilities, style issues, or potential bugs), but they do not satisfy the project's review requirements.

> * **Humans make the decisions.** Maintainers and component owners make merge decisions following the project's [governance framework](./GOVERNANCE.md#pr-approval-rules). PR approval requires votes from maintainers and/or component owners as defined in the governance policy.

Reviewer 2: ```suggestion
* **Humans make the decisions.** Maintainers and component owners make merge decisions following the project's [governance framework](./GOVERNANCE.md#pr-approval-rules). PR approval requires votes from maintainers, component owners, or both as defined in the governance policy.
```

* **Contributors should engage with human reviewers.** While you may find AI-assisted review tools helpful during development, your PR must be reviewed and approved by project maintainers or component owners before merging.

## AI-Assisted Strimzi Proposals

[Strimzi proposals](https://github.com/strimzi/proposals) are strategic documents that significantly impact users or project direction. Due to their importance, special considerations apply:

* **AI-assisted proposals require deep understanding.** If you use AI tools to help draft a proposal, you must have a thorough understanding of the problem space, proposed solution, alternatives considered, and implications for users and the project.

> * **The rationale matters most.** Proposals are not just about "what" to do, but "why" it should be done, what alternatives exist, and what trade-offs are being made. AI tools may help structure these thoughts, but the reasoning must come from genuine understanding of the project and its users.

Reviewer 2: ```suggestion
* **The rationale matters most.** Proposals are not just about "what" to do, but "why" it should be done, what alternatives exist, and what trade-offs are being made. AI tools may help structure these ideas, but the reasoning must come from genuine understanding of the project and its users.
```

* **Disclose AI usage.** If AI tools play a significant role in drafting a proposal, note this in the PR description.

> * **Engage authentically in discussions.** Proposal discussions involve community consensus-building and voting. You must engage directly with the community, not through AI-generated responses. Maintainers and community members need to understand your reasoning and evaluate the proposal based on your understanding, not an AI's output.

Reviewer 2: ```suggestion
* **Engage authentically in discussions.** Proposal discussions involve community consensus-building and voting. You must engage directly with the community, not through AI-generated responses. Maintainers and community members need to understand your reasoning and evaluate the proposal based on your understanding, not AI-generated output.
```

* **AI-assisted reviews of proposals.** You may use AI tools to help you understand and evaluate proposals, but your vote or feedback must reflect your own judgment and reasoning.

> Proposals that appear to be largely AI-generated without genuine understanding of the problem space MAY be closed by the maintainers.

Reviewer 2: ```suggestion
Proposals that appear to be largely AI-generated without genuine understanding of the problem space will be closed by maintainers.
```

Reviewer 2: we would just shut it down, no?

Author: I would say yes :-)

## Questions

> If you have questions about this policy or are unsure whether your use of AI tools requires disclosure, please contact the maintainers.

Reviewer 2: Should AI-generated code or content be explicitly flagged in comments

Author: Are you referring to comments when engaging with maintainers during a PR review? Isn't it too much adding on each comment that it was generated via AI? We already said that you should mention you used AI in the PR description, maybe is it enough?

Reviewer 2: I meant comments in the code

# AI Contribution Policy for the Strimzi Project

## Overview

Contributors may use AI tools, such as Large Language Models (LLMs) and code assistants, when contributing to Strimzi.
As with any development tool, the contributor is responsible for the quality of the result and for understanding what they submit.

**You are the contributor.** When you submit a Pull Request (PR) to Strimzi, you certify the contribution as your own work.
AI-generated or AI-assisted content does not change this responsibility.

## Requirements

### Understanding and Responsibility

* **You must understand your contribution.** Do not submit code, documentation, or other content that you do not fully understand. You should be able to explain to reviewers what your contribution does, how it works, and how you verified it.
* **Review all AI-generated content.** Before submitting, verify that the contribution is correct, tested, and follows the project's contributing guidelines.

> * **Use AI as a learning tool.** New contributors could use AI tools to understand the codebase, patterns, and best practices. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.

Reviewer 4: Maybe @PaulRMellor should confirm, but I think this should be `can`

```suggestion
* **Use AI as a learning tool.** New contributors can use AI tools to understand the codebase, patterns, and best practices. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.
```

* **Use AI as a learning tool.** New contributors can use AI tools to understand the codebase. If you're using AI to generate code you don't understand, take time to learn the concepts before submitting.
* **You own what you submit.** As the PR author, you are responsible for all changes, regardless of the tools used to create them.

### Disclosure

* **Disclose AI usage in your PR description.** If you used AI tools in preparing your PR, you must disclose this in the description of your PR. Listing AI tooling as a co-author, co-signing commits using an AI tool, or using the `Assisted-by`, `Co-authored-by`, `Co-developed-by`, or similar commit trailer is not allowed.

### What NOT to Do

* **Do not use AI tools in commit trailers.** AI tools MUST NOT be added to `Signed-off-by`, `Co-authored-by`, `Assisted-by`, `Co-developed-by`, or similar tags in commit messages. Only humans can legally certify the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). The `Signed-off-by` tag certifies that you wrote the code or otherwise have the right to submit it under the project's license. An AI tool is not a legal entity and cannot make this certification.

> * **Engage authentically with reviewers.** When maintainers, component owners, or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated responses. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.

Reviewer 4: Since this is under the "What NOT to Do" section I've proposed some tweaks so all the headings are things not to do
```suggestion
* **Do not instruct AI tools to directly respond to reviewers.** When maintainers, component owners, or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated responses. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.
```

> * **Do not instruct AI tools to directly respond to.** When maintainers, component owners, or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated responses. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.

Reviewer 4: ```suggestion
* **Do not instruct AI tools to directly respond to comments.** When maintainers, component owners, or other contributors provide feedback on your PR, you may use AI tools to help craft responses, but you must fully understand and take ownership of what you write. Reviewers want to engage with you, not receive generic AI-generated responses. If you use AI assistance to formulate responses, ensure they genuinely reflect your understanding and that you can defend your reasoning.
```

> * **Do not submit large AI-generated PRs.** AI tools can generate content faster than reviewers can read it. Keep PRs focused and appropriately sized. Strimzi maintainers MAY close pull requests that are too large to review effectively or where the contributor does not appear to understand the changes.

Reviewer: ```suggestion
* **Do not submit large AI-generated PRs.**
    * AI tools can generate content faster than reviewers can read it. Keep PRs focused and appropriately sized. Strimzi maintainers MAY close pull requests that are too large to review effectively or where the contributor does not appear to understand the changes.
    * Before opening large PRs with major changes, consider reaching out to the maintainers first, to better understand whether the project is interested in such a contribution in the first place.
```

Reviewer 5: ```
Before opening large PRs with major changes, consider reaching out to the maintainers first, to better understand whether the project is interested in such a contribution in the first place.
```
This seems like a useful suggestion to add elsewhere like in CONTRIBUTING as it's not just about using AI?

Reviewer: In general, you are right, it is valid everywhere. But outside of the AI realm, the effort is at least symmetric on both sides. With AI, the effort becomes very asymmetric as users can create huge changes with a single comment. That is why I think it is even more important here.

Author: I agree with @tinaselenge, that suggestion is useful in general and not AI related. Of course, it's true that with AI the possibilities to get big PRs are higher than before. Here we are focusing on "Do not submit large AI-generated PRs.", so our expectations is that we don't get such PRs at all. Adding that "before opening large PRs ... etc etc" seems a contradiction. So again I am happy with that suggestion as useful one but not here.

Reviewer: You are right that this suggestion is general and not strictly AI related. But when did we last got unwanted really large PR written by hand?

In any case, should we then remove the whole point? I was trying to rephrase it as _ask first and let's try to figure it out_. But what we have now is forbidding large AI based PRs even if suitable, agreed on, and desired.

> * **Focus on quality, not quantity.** Do not open multiple AI-generated PRs in rapid succession. Submitting many PRs at once or within a few minutes overwhelms reviewers and will result in PRs being closed.

Reviewer 4: ```suggestion
* **Do not open multiple AI-generated PRs in rapid succession.** Submitting many PRs at once or within a few minutes overwhelms reviewers and will result in PRs being closed. Instead, focus on quality over quantity.
```

> * **Do not open multiple AI-generated PRs in rapid succession.** Submitting many PRs at once or within a few minutes overwhelms reviewers and will result in PRs being closed. Instead, focus on quality over quantity.

Reviewer: Maybe many would be better? You use it in the follow up sentence as well.

```suggestion
* **Do not open many AI-generated PRs in rapid succession.** Submitting many PRs at once or within a few minutes overwhelms reviewers and will result in PRs being closed. Instead, focus on quality over quantity.
```

* **The rationale matters most.** Proposals are not just about "what" to do, but "why" it should be done, what alternatives exist, and what trade-offs are being made. AI tools may help structure these ideas, but the reasoning must come from genuine understanding of the project and its users.
* **Disclose AI usage.** If AI tools play a significant role in drafting a proposal, note this in the PR description.
* **Engage authentically in discussions.** Proposal discussions involve community consensus-building and voting. You must engage directly with the community, not through AI-generated responses. Maintainers and community members need to understand your reasoning and evaluate the proposal based on your understanding, not an AI-generated output.
* **AI-assisted reviews of proposals.** You may use AI tools to help you understand and evaluate proposals, but your vote or feedback must reflect your own judgment and reasoning.

Proposals that appear to be largely AI-generated without genuine understanding of the problem space will be closed by maintainers.

## Questions

> If you have questions about this policy or are unsure whether your use of AI tools requires disclosure, please contact the maintainers.

Reviewer 4: I guess we should say where to contact us
```suggestion
If you have questions about this policy or are unsure whether your use of AI tools requires disclosure, please contact the maintainers via Slack or the mailing list.
```
